### DISCUSSION TITLE
How to accelerate the algorithm to find cycles in a DAG?

### DISCUSSION THREAD

**Discussion by Neronjust2017** [networkx/networkx#5724] [Q&A] [answered]
Hi, I want to know if I can use multi-threading or something else to accelerate the cycle finding functions, such as recursive_simple_cycles ?

**Comment by dschult** [comment, marked as answer]
threading and recursion are like oil and water right?  Also, a DAG should not have any cycles (hence acyclic), but I guess you mean cycles in a DiGraph.

Even a nonrecursive version of the function would still use a queue that holds a state, so not good with multi-threading.
Maybe you can find a way to split the problem up into smaller problems that are independent. But I can't think of any general approach.

**Reply by rossbar** [reply]
Agreed - at the very least, the cycle-finding function is not "naively" or "embarrassingly" paralleli. Any ideas or references on how certain functions can be parallelized are of interest - I'd encourage anyone interested to consider adding to the discussion in #4064.

**Comment by Neronjust2017** [comment]
Sorry for that. Yes, I mean&nbsp;DiGraph. Thanks!



发自我的iPhone


------------------ Original ------------------
From: Dan Schult ***@***.***&gt;
Date: Sun,Jun 12,2022 1:24 AM
To: networkx/networkx ***@***.***&gt;
Cc: Neronjust2017 ***@***.***&gt;, Author ***@***.***&gt;
Subject: Re: [networkx/networkx] How to accelerate the algorithm to find cycles in a DAG? (Discussion #5724)





 
threading and recursion are like oil and water right?  Also, a DAG should not have any cycles (hence acyclic), but I guess you mean cycles in a DiGraph.
 
Even a nonrecursive version of the function would still use a queue that holds a state, so not good with multi-threading.
 Maybe you can find a way to split the problem up into smaller problems that are independent. But I can't think of any general approach.
